New stat parse

sergpolly opened the pull request:
thoroughly tested commit that introduces `stats` to `parse`, using a `class StatObject`(name is bad, I agree). Such an implementation does not degrade the performance in a noticeable way, and allows us to (1) simplify the code for `pairsamtools stats` themselves (2) introduce `stats` gathering into `pairsamtools parse` in a least obtrusive way(not very obtrusive, let's put it this way).

I've tested the reimplemented `stats` (including merge) on test data and in a big data scenario, where I merged 13 chunks of `.stats` files into one. Stats match perfectly with previous implementation, except for the `total_single_sided_mapped`, which was initialized with 1 instead of 0 for some reason. I fixed it to`stats['total_single_sided_mapped'] = 0`.

Let me know if an implementation using a `class` is not desirable, and please feel free to rename `class methods`, which I dislike greatly, but cannot come up with anything better. Complementary distiller PR is coming right away.

sergpolly commented:
build is failing because there is a wrong number of `total_single_sided_mapped` in the assertion, it should be 3 instead of 4, check input:
https://github.com/mirnylab/pairsamtools/blob/master/tests/data/mock.pairsam

I can count only 3 `total_single_sided_mapped` - do I misinterpret something?

nvictus commented:
Looks good! Though, the `Methods` subsection is probably overkill. Sphinx autodoc can take care of that when we eventually add sphinx docs.

sergpolly commented:
S0 - acepting as is? or waiting for the nested dict/parsing the file solution ?

nvictus commented:
I'm happy accepting as is. @golobor?

sergpolly commented:
I've implemented `StatObject` that would parse `.stats` file properly and return the same type of object as you would expect from building up `StatObject` from scratch (empty `init` and then `update` with pairs information).

I had to imply a rigid structure of the `OrderedDict` underlying `StatObject`, i.e., while parsing it explicitly implies that the nested sub-dicts should be names `pair_types`, `chrom...` and `dist_freq` and parses each category differently. `dist_freq` in particular, since the corresponding distribution is stored in `ndarray`. I've tried to make use of exceptions to the best of my knowledge - mostly copying your patterns of exceptions usage. 

I also had to make `from_file` a factory method or `@classmethod` for it become another constructor.

I also implemented nested indexing -  the [flatdict](https://github.com/gmr/flatdict) style: "pair_types/CX"  is a valid key, so you can say: `s = StatObject.from_file('in.stats')` and then `s['pairs_type/CX']`. Why do we need this I don't know, but it's there and it's consistent with the structure of a `.stats` file
